### DISCUSSION TITLE
REDESIGN: Model Animation System

### DISCUSSION THREAD

**Discussion by JeffM2501** [raysan5/raylib#4606] [Ideas]
# Point of Discussion
There has been some discussion about what we can do to make raylib's 3d animation system more robust. The current API and model structure are heavily focused around CPU based animation and there being one animation instance per model. With the addition of GPU animation support and advanced needs of raylib users, this API is proving to be weak.

Several people have different ideas on how to enhance the API, but in different ways, and this leads to confusion and large PR "code drops" that people don't understand.
This discussion is to talk about the needs of raylib and come up with a design before talk of any code.

# Design of raylib
Raylib is designed first and foremost as a library to help people get into gamedev and learn how lower level systems are made. As has been said many times before, raylib is not an engine, and along that line, we should not entertain complex systems that are not easy for new users to pickup and user. Raylib APIs are meant to be simple and meant to be extended. We should focus on the lowest level APIs possible to do what is desired, and leave higher level systems up to external/drop in code (such as what is hosted on raylib-extras).

# GPU vs CPU animation
Raylib has traditionally only supported CPU based animation, with GPU support just added in 5.5, because of this much of the API and structure is based around the model and mesh structure storing the current state of the transformed vertices. This is not always the best case for GPU based animation.

# Splitting the current pose out from the model
Currently the animation data and animation pose cache is stored spread over several areas, some in the model, some in the animation, and some in the mesh.
I don't feel that storing any animation data in the mesh is correct, as there are no raylib APIs that use the data directly on the mesh, it's always used in the context of a model. Thus I propose that we remove the animation data from the mesh itself, `DrawMesh `doesn't need it, so why should it be there?

We should make a new structure called `ModelAnimationPose`, and it should contain a list of MeshAnimationPoses, in the same order as the meshes in a model. This structure would contain the animation related data that is currently in the mesh, most importantly the `boneMatrices `list. The `boneMatrices` is a cache of the current bone transforms for a pose. It is what the base vertices are transformed by for CPU animation, and are what is uploaded to the shader for GPU animation.

The model structure itself should hold one of these pose structures, giving the model the same storage for a 'current pose' that it has now, allowing all the existing animation functions to continue to work.

Any new APIs that change the pose on a model would take a `ModelAnimationPose `as input and modify it to the new post. Again, this lets all existing animation functions work as expected.

We should make low level functions that apply a type of animation to a pose, such as
```SetModelAnimationPoseToFrame(Model*, ModelAnimation*, ModelAnimationPose*, int frame);```
```InterpolateModelAnimationPoseToFrame(Model*, ModelAnimation*, ModelAnimationPose*, int frame1, int frame2, float paramater);```
These functions would simply update the pose passed in.

GPU based animations would then just upload the matrix cache for a mesh from the post as they draw the model.

For CPU animation we should have a function that applies a pose to the vertices of a model, by transforming the verts and uploading them to the GPU
```ApplyPoseToModel(Model*, ModelAnimationPose*);```

# Pose Per Instance
Now that we have the animation system working on a self contained pose structure, we can add functions to extend this to add more features.

We can add a function to clone a pose.
```ModelAnimationPose CopyAnimationPose(ModelAnimationPose*);```
This will allow games that reuse the same model with different animation states to cache a pose per instance and update them as necessary. Right now if I have 10 copies of a mesh in different parts of the animation, I have to constantly rewrite the boneMatrices of the meshes before I draw anything.  I would much rather maintain my own cache of poses and update them as needed (possibly in a thread), but the current system prevents that.

We can then also look into adding additional functions, internal, or external to update a pose in different ways such as doing animation blending, or programmatic modification of bones for features such as mouse control or inverse kinematics.

The separation of pose and model gives us a large range of possibilities for the future, with no real downside on the raylib side.

# Scope
I think we should start with this low level of changes so that we can have a good foundation for storing and applying animation states as poses. This will then people expand as needed. 
These low level functions will be used to reimplement the current high level animation functions so they work the same as they always have.

Then we can prototype/discuss further high level functions after this low level framework is in place.

**Comment by Kirandeep-Singh-Khehra** [comment]
IMO Using `ModelAnimationPose` as name might not be a good idea. Because across raylib the term **"pose"** is generally represented by `Transform *`. Like `bindPose` in `Model` is `Transform *` and `framePoses` in `ModelAnimation` is of `Transform **`(Array of poses). So, it might create two definitions of word `Pose`. I would suggest words similar to `Rig`(not good name in animation domain), `Skeleton`(way too broad) or `Armature`(good but not used a lot in game development but in sounds good in modeling) (<-- Please don't consider these as suggestions for functionality but name and a rough base for new Name).

1. `Model.bindPose` reference
https://github.com/raysan5/raylib/blob/79facde3535af1f35942eec38f02f2851de618ab/src/raylib.h#L417
2. `ModelAnimation.framePoses` reference https://github.com/raysan5/raylib/blob/79facde3535af1f35942eec38f02f2851de618ab/src/raylib.h#L425

**Comment by Kirandeep-Singh-Khehra** [comment]
As per the requirements discussed i am trying to implement a drop in system here https://github.com/Kirandeep-Singh-Khehra/raylib-3d-anim-system. It is using layer based animation approach with `Skeleton` and `Pose` and needs no change in raylib code. Here are types defined:

```c
typedef Transform* Pose;

typedef struct Skeleton {
  /* Extracted from Model */
  int boneCount;         // Number of bones
  BoneInfo *bones;       // Bones information (skeleton)
  Pose bindPose;   // Bones base transformation (pose)

  /* Extracted from Mesh */
  Matrix *boneMatrices;  // Bones animated transformation matrices (not used yet)

  /* Custom defined */
  Pose pose;      // Current pose
} Skeleton;
```

And here is a basic usage example.
https://github.com/Kirandeep-Singh-Khehra/raylib-3d-anim-system/blob/main/examples/skeleton/skeleton_animation_blending.c

Primary intent for this repo is to act as a playground or external base and then add some contributions (step by step) of basic commonly used and simple functions to raylib when it will be good enough to call it stable. Currently naming convention is poor and some redundant functions are present.

Please do give it a look(and little review if possible). And if anything in it looks good and can be added safely to raylib now or anytime in future then do let me know.

**Reply by Kirandeep-Singh-Khehra** [reply]
We might also need a pose with local transformations(relative to parent bone). It can add possibility of Forward Kinematics, better blending and additive blending. and it will make it little easy to have IK(Inverse Kinematics) too.

**Reply by JeffM2501** [reply]
I highly recommend we get the foundational API in place before we work on drop in libs for higher level functions.

**Reply by Kirandeep-Singh-Khehra** [reply]
Yes, i totally agree, and i am on same path but different approach. I m trying to implement high level stuff and then breaking down those to build a low level functions and then iterating back up. It also helps me to get an idea of what all is needed. And as per my current progress only `Transform* pose` is required in `struct Model` to do all things related to pose. And some bunch of functions around `typedef Transform *Pose`. This makes it easier to work with `anim.framePoses[frame]` too. Like some implemented [here in pose.c](https://github.com/Kirandeep-Singh-Khehra/raylib-3d-anim-system/blob/main/src/pose.c). But this implementation is not good in terms of memory management and there are some useless functions too. I need to rewrite this whole part(I know i'm again jumping to code. But that's my habit to continuously iterate up and down and i'm not forcing any high level functions but just trying to get the minimal set and structure of low level systems for easy implementation most of high level things).

**Comment by JeffM2501** [comment]
In looking at the current API, one thing that I don't like is that DrawMesh sets the uniforms in the shader, I think that should be done at model/material level.
Right now we are re-sending all the bones for all meshes for every mesh.
I think DrawModel should walk the material list and upload all the matrices once, from the pose.
this way we also don't need to deep copy the pose to every mesh.

**Comment by JeffM2501** [comment]
I have made a prototype branch with some low level functions to support the ideas I proposed above.
https://github.com/JeffM2501/raylib/tree/external_animation_pose

The main change is moving the animation support out of the mesh and into the model. The bones are already all the bones for all meshes in a model, and an animation is model based.

All the existing APIs work exactly as they did before, the animation data is just now stored in a "ModelBonePose" structure on the model instead of duplicated for each mesh, thus this change uses less memory and does fewer deep copies than the previous one, but the model structure is now 1 size_t larger, while the models are 1 int and 1 size_t smaller.

I have also added a few extended APIs that work directly on ModelBonePose structures to allow callers to keep separate states for multiple draws of the same model during the same frame. The GPU mesh animation example has been updated to run 2 animations to show this.

The only breaking change is that this API no longer supports animation during DrawMeshInstanced using a GPU animation shader, because the animation data is not part of the mesh anymore. If we want to support animation in instanced models, then we'd need to add DrawModelInstanced that did the matrix upload. Note that I don't think anyone did this before now (all instances would need to use the same pose), so I don't think anyone is majorly broken.

If this looks good, it can be extended to add interpolation support when updating a pose. I am not married to any of the specific names, so if there are better suggestions, please let me know.

![anim_2](https://github.com/user-attachments/assets/eab2046b-0b77-4cd8-81b1-ecedd3dd45a4)

**Reply by Kirandeep-Singh-Khehra** [reply]
This is really good and easy to understand. Working with matrices for pose is a great idea. And a lot less intermediary steps between loading, updating and draw calls.

**Reply by Kirandeep-Singh-Khehra** [reply]
I don't know most of maths but will matrices(as pose) cause some issues like buggy interpolation(considering matrix decomposition as overhead to fix it) and other things like position of bone attachments(Only idea in my mind is to first move the object/attachment(like sword) at place with respect to bindpose using transform matrices and then multiply it to its parent bone matrices to keep it in place). And advanced foot placement and collisions in ragdolls(and other examples which need position of bone or only one component of transform like position, rotation or scale) can become complex. Are their any ways or plans to make this easy and simple.

**Comment by JettMonstersGoBoom** [comment]
This is a good proposal, as folks try and create more advanced animation systems for their games, By allowing the developer to update the matrices as they see fit is a really positive step. 
This change doesn't need to be the most complex system, and is actually simpler than the current system to implement ( once untangled ) 
the previous animation system was too heavily tied to the model structures and as has been pointed out, led to confusion. 
this change opens the door to ragdolls, IK, FK, and also more mesh and animation formats. 
Is there any thought to supplying an API that matches the current one, using the new system with appropriate setup ?

**Comment by JeffM2501** [comment]
I have found some more duplicated data that I am in the process of consolidating,
1)each animation has a full copy of the bone info for all bones, this seems superfluous. Every time we use the animation we have a model since we need the bind pose, so it makes sense to store the skeleton entirely in the model.

I am going to make some incremental changes
1) move all bone info into the model and just keep one copy.
2) see about moving the entire skeleton into it's own structure just to keep it logically grouped.

**Reply by JeffM2501** [reply]
I guess the only real reason the animation has any info about the bones is that it's used to validate that the animation goes with the model.

This seems like a lot of data just to validate something the user has to manage anyway.

Should it just not be on the user to ensure that the animations go with the skeleton? Raylib doesn't do this kind of compatibility check for anything else.

**Reply by JettMonstersGoBoom** [reply]
Minimum expectation is throw a log error with mismatched number of bones and refuse to update at all. The model only needs it's bind pose but keeping bonenames could be useful AND help detect mismatched data ?
I agree it's on the user to use the correct skeleton though.

**Reply by JeffM2501** [reply]
I don't want to remove data from the model, I want to remove the duplicate bones from the each animation sequence is storing a complete copy of the boneinfos, 36 bytes per bone, per sequence, all duplicated.

I'm trying to think where else raylib logs an error if the user screws up like this. If you pass in the wrong image buffer size to a raylib function, it does not log an error, it just crashes.

**Reply by JeffM2501** [reply]
Optimally we could compute a hash for each unique skeleton and just store that for error checking, but that may be overkill.
Right now any validation of model->animation binding involves checking every bone to ensure they have the same parent linkage, names are not used for validation.

**Comment by JeffM2501** [comment]
To continue, I have also implemented a C++ system for my personal project that basically does what my proposed change to raylib does (just with some C++ features for my specific needs).

it's all part of this little 'model_lib'.

https://github.com/JeffM2501/ModernBoomerShooter/blob/main/model_lib/include/model.h

**Comment by Kirandeep-Singh-Khehra** [comment]
I think while making these changes we can also take care of `ModelAnimation` and add support for variable frame rate or load animation poses as given in input file. With one additional float parameter to store frame delay. Currently we can load it as it is and set that frame delay value to 1/60 and worry about loading afterwards.

**Comment by raysan5** [comment]
@JeffM2501 @Kirandeep-Singh-Khehra @JettMonstersGoBoom Hi! Finally got some time to carefully go through this proposal. Thank you very much for all the thinking already put on it, most of the proposed changes look great to me and seem simple enough to not break the API (only the model instancing functionality, but not dramatic).

Sincerely, the harder part for me at this point is, as usual, in the proper naming of structures and variables.

@Kirandeep-Singh-Khehra your `Skeleton` approach was actually extremely similar to my first approach for animations, just simplified it later to minimize number of structures.

@JeffM2501 About CPU vs GPU animation, I'm considering a build flag to select one of both (afair Unity had a similar approach), it will simplify the API for users, with no need to worry about the function to use and could also help with internal "fast-code" paths.

**Comment by aeaeaeaeaeae** [comment]
I have a consideration. I think Skeletons should be though of as a primitive of their own. They are not intrinsically bound to either the mesh or the model. A `skeleton` (a set of bones) makes up the state of a rig (partially). Charachter effects, such as FK, IK or other types of procedural animation operate on the **skeleton**, which then later might or might not be applied to a mesh. 

Having that data somewhat independent of the model would open a lot of unforseen usecases. It's also cleaner imo to pass around a Skeleton to a potential future character effect API, as a Model contain a lot of irrelevant data. You might not even want to apply the final pose to a model at all in the end.

I think a `BonePose` (rather than `ModelBonePose` and `Pose`) is a good term for a state of a skeleton, where `binePose` is the pose when binding the mesh to the skeleton, and `framePose` a pose on a given frame in the context of a cached animation. Makes sense.

```C
typedef struct BonePose {
	Matrix* boneMatrices;   // Bones transform matrices
	int boneCount;          // Number of bones
} BonePose;

typedef struct Skeleton {
	BonePose bindPose;  // or restPose to make it even independet of the mesh
	BonePose activePose;
	BoneInfo *info;
        int boneCount;
} Skeleton;

typedef struct SkeletonAnimation {
	BonePose *framePoses;
} SkeletonAnimation;

typedef struct Model {
	...

	Skeleton skeleton;

	// The animation should probably not even be an intrinsic part of the model,
	// as you usually have multiple animation applied to the same model,
	// and since animations are also loaded independently of the model the  user
        // has to manage this themselves anyways.
	SkeletonAnimation anim;

} Model
```

**Comment by TokyoSU** [comment]
An idea im thinking using assimp to load model using a tool like rlModelConverter which convert any model format to raylib format would be nice, we wont need to have external library that load just 1 model type, removing these and using assimp to load and export it to something like "model_name.rlmodel" would be more clean to do.
That would allow raylib to support all assimp format since assimp have a uniform api when loading models:
https://github.com/assimp/assimp/blob/master/doc/Fileformats.md
And if the model also contains animations, we could also export them to something like "model_name@run.rlanim".
Or all of them inside the same file: "model_name.rlanim"
Using command-line or gui to customize export if required.

**Reply by TokyoSU** [reply]
Or having a plugin inside blender which export to raylib format directly would be nice like how m3d do.

**Reply by ElectroGamesDev** [reply]
Raylib doesn't have its own format. GLTF/GLB would be the best format to use which is supported by both Blender and Raylib. If you were suggesting creating a new format, there really isn't much point as GLTF is already open-sourced, easy to develop for, and mature. ASSIMP support would be cool, but its way too big of a dependency so Ray wouldn't allow it to be built in by Raylib. Not to mention it would be a lot of work and ASSIMP is known to have issues which is why Godot dropped support for it.

**Reply by TokyoSU** [reply]
That would be an external tool just dropping a model on it to convert it, raylib would have only the model loader, well if assimp have issues then don't mind the message above :x

**Reply by dc1248** [reply]
Sure, you could make your own custom format and loader, I've done that before Raylib and might do it again if needed for optimized model loading. It doesn't need to be part of Raylib though. I could argue that it should be, because this format would match Raylib's model/mesh/animation structures - but it should start as an experimental add-on.

GLTF works fine for me using these options in the Blender exporter:
- export models in .gltf text format and enable 'Keep original images'
- export animations separately in .glb binary format

**Comment by raysan5** [comment]
@JeffM2501 @Kirandeep-Singh-Khehra @aeaeaeaeaeae Excuse the late response. I'm finally reviewing all the proposals for a more flexible and comprehensive animation system for raylib, with not too-much-breaking changes and still keeping the overall simplicity.

Following your ideas, I propose the following data structures redesign:

```c
// Mesh vertex data
typedef struct Mesh {
    int vertexCount;            // Number of vertices stored in arrays
    int triangleCount;          // Number of triangles stored (indexed or not)

    // Vertex attributes data
    float *vertices;            // Vertex position (XYZ - 3 components per vertex) (shader-location = 0)
    float *texcoords;           // Vertex texture coordinates (UV - 2 components per vertex) (shader-location = 1)
    float *texcoords2;          // Vertex texture second coordinates (UV - 2 components per vertex) (shader-location = 5)
    float *normals;             // Vertex normals (XYZ - 3 components per vertex) (shader-location = 2)
    float *tangents;            // Vertex tangents (XYZW - 4 components per vertex) (shader-location = 4)
    unsigned char *colors;      // Vertex colors (RGBA - 4 components per vertex) (shader-location = 3)
    unsigned short *indices;    // Vertex indices (in case vertex data comes indexed)

    // Animation vertex data
    //------------------------------------------------------------------------------------------
    // Skin data for the current mesh (MAX: 256 bones supported)
    int boneCount;              // Number of bones affecting Mesh
    unsigned char *boneIndices; // Vertex bone indices, up to 4 bones influence by vertex (skinning) (shader-location = 6)
    float *boneWeights;         // Vertex bone weight, up to 4 bones influence by vertex (skinning) (shader-location = 7)

    // Cache CPU animation vertex data
    float *animVertices;        // Animated vertex positions (after bones transformations)
    float *animNormals;         // Animated normals (after bones transformations)

    // Runtime animation data kept out of Mesh (as much as possible)
    //------------------------------------------------------------------------------------------

    // OpenGL identifiers
    unsigned int vaoId;         // OpenGL Vertex Array Object id
    unsigned int *vboId;        // OpenGL Vertex Buffer Objects id (default vertex data)
} Mesh;

// Bone, skeletal animation bone
typedef struct BoneInfo {
    char name[32];              // Bone name
    int parent;                 // Bone parent
} BoneInfo;

// Typedef for convenience (just in case, may be not needed)
typedef Transform *AnimPose;

// Skeleton, defines model skeleton
typedef struct Skeleton {
    int boneCount;
    BoneInfo *bones;            // Bones info
    AnimPose bindPose;          // Bind pose (local)
} Skeleton;

// Model, meshes, materials and animation data
typedef struct Model {
    Matrix transform;           // Local transform matrix

    int meshCount;              // Number of meshes
    int materialCount;          // Number of materials
    Mesh *meshes;               // Meshes array
    Material *materials;        // Materials array
    int *meshMaterial;          // Mesh material number

    // Animation data
    Skeleton *skeleton;         // Skeleton used by model

    // Runtime animation data (to be updated for drawing)
    int boneCount;              // Really required? Or we can use skeleton provided info?
    AnimPose currentPose;       // Current animated pose (really required?)
    Matrix *boneMatrices;       // Current animated transformation matrices
} Model;

// ModelAnimation
typedef struct ModelAnimation {
    char name[32];              // Animation name

    Skeleton *skeleton;         // Skeleton for the animation (must match Model one to apply animation)

    int framePoseCount;         // Number of animation frames
    AnimPose *framePoses;       // Poses array by keyframe
} ModelAnimation;
```

With this new design different `Models` can share a `Skeleton` and every `Model` has its own animation current state depending on the `ModelAnimation` frame applied to them.

`ModelAnimations` can be loaded independently of `Mesh`/`Model` data and blending can be compute independently at `ModelAnimation` level as proposed before being applied per `Model`.

Please, let me know your thoughts.

References:

 - [`UpdateModelAnimationBonesWithBlending()`](https://github.com/Kirandeep-Singh-Khehra/raylib/commit/8808a3958a705120bea516cf8d50155e5b9d8725)
 - [external-animation-pose](https://github.com/JeffM2501/raylib/blob/external_animation_pose/src/raylib.h)
 - [raylib-3d-anim-system](https://github.com/Kirandeep-Singh-Khehra/raylib-3d-anim-system)
 - [`UpdateModelAnimationBonesLerp()`](https://github.com/raysan5/raylib/pull/4578/changes#diff-008ec254b9e10087d979356eaa799f93f24b3173a6a0c44445c24fee791a1e90)

**Reply by JeffM2501** [reply]
I think that will help a lot

**Reply by JeffM2501** [reply]
The only reason to keep a bone count in the model is to compare it to the skeleton to verify that it will work. Optimally this would be a hash of the bone names/connections.

The current pose does not need to be stored as long as there is a function to build up a set of bone matrices from any arbitrary pose, then interpolation is possible

**Reply by raysan5** [reply]
> The current pose does not need to be stored as long as there is a function to build up a set of bone matrices from any arbitrary pose, then interpolation is possible

Yeah, that was a doubt, if it could be useful to keep it for some reason, along `Matrix *boneMatrices`, but they are redundant.

**Reply by Kirandeep-Singh-Khehra** [reply]
I was doing interpolation on `Transform*` like this.
https://github.com/Kirandeep-Singh-Khehra/raylib-3d-anim-system/blob/07ef46fd4b1eee2094aa5d070f13cc4beb261539/src/pose.h#L76
```c
typedef Transform *Pose;

Pose PoseLerp(Pose poseA, Pose poseB, int boneCount, float factor) {
  Pose pose = InitPose(boneCount);

  for (int i = 0; i < boneCount; i++) {
    pose[i] = TransformLerp(poseA[i], poseB[i], factor);
  }

  return pose;
}
```

And for lerping/transitioning from current pose to target pose(lagging animation) will need currentPose.

And bevy and more engine/framework i referred does blending with Transforms.

I am assuming that plan is to use Transform* for computing blend and not matrices. And using matrices to blend can also warp model(not sure).

**Reply by raysan5** [reply]
Thanks, it makes sense, we can keep it for blending.

**Comment by raysan5** [comment]
@JeffM2501 @Kirandeep-Singh-Khehra @JettMonstersGoBoom I keep working on this redesign, here it is the current proposal with some concerns following...

```c
// Transform, vertex transformation data
typedef struct Transform {
    Vector3 translation;    // Translation
    Quaternion rotation;    // Rotation
    Vector3 scale;          // Scale
} Transform;

// Anim pose, an array of Transform[]
typedef Transform *AnimPose;

// Bone, skeletal animation bone
typedef struct BoneInfo {
    char name[32];          // Bone name
    int parent;             // Bone parent
} BoneInfo;

// Skeleton, animation bones hierarchy
typedef struct ModelSkeleton {
    int boneCount;          // Number of bones
    BoneInfo *bones;        // Bones information (skeleton)
    AnimPose bindPose;      // Bones base transformation (Transform[])
} ModelSkeleton;

// Model, meshes, materials and animation data
typedef struct Model {
    Matrix transform;       // Local transform matrix

    int meshCount;          // Number of meshes
    int materialCount;      // Number of materials
    Mesh *meshes;           // Meshes array
    Material *materials;    // Materials array
    int *meshMaterial;      // Mesh material number

    // Animation data
    ModelSkeleton skeleton; // Skeleton for animation

    // Runtime animation data (CPU/GPU skinning)
    AnimPose currentPose;   // Current animation pose (Transform[])
    Matrix *boneMatrices;   // Bones animated transformation matrices
    int boneMatricesLoc;    // Shader bone matrices location
} Model;

// ModelAnimation, contains a full animation sequence (AnimPoses[])
typedef struct ModelAnimation {
    char name[32];          // Animation name
    int boneCount;          // Number of bones
    int keyframeCount;      // Number of animation key frames
    AnimPose *keyframePoses; // Animation sequence keyframe poses [keyframe][AnimPose]
} ModelAnimation;
```
Concerns:
1. ~I defined `AnimPose` as `typedef Transform *AnimPose;` for convenience but not sure if it will be clearer for users to just use `Transform *pose` instead. `AnimPose` "hides" an array of data.... or maybe it's just fine for most of users?~
2. ~`ModelSkeleton` contains `bindPose` but I noticed `LoadModelAnimations()` does not initialize that `bindPose`, should it be loaded from model file? should it be copied (referenced) from model on animation application? should it be just left NULL (as I do now)?~
3. ~Despite `Model.skeleton` and `ModelAnimation.skeleton` can potentially be shared, at the moment two copies are loaded. It could be used for validation but not sure if it is the best approach to keep two copies, thoughts?~
4. In this new design, we have `Model.currentPose` that at this moment is not used, should it be initialized with some data on `Model` loading? I think `UpdateModelAnimation()` should fill it, but it will probably imply some redesign to current design.
5. ~Redesigned `boneMatrices` shader setup, instead of trying to get location internally, I'm exposing now `Model.boneMatricesLoc`. What do you think? Should it be kept internal like it was? I don't specially like to keep a `boneMatricesLoc` around but trying to set it automatically neither seems that right... I don't know...~
6. I reviewed all the code but I'm not sure if I'm missing something on how IQM/M3D/GLTF models and model-animations are loaded, I'd need some help on that regards.

I'm planning to merge all changes as soon as possible. It could require reviewing all involved examples and adding some new functions for AnimPoses blending. @Kirandeep-Singh-Khehra once this change meerged, could you review your PR for animation blending support?

**Reply by raysan5** [reply]
`ModelSkeleton` is effectively not needed on `ModelAnimation`, removed it.

**Reply by raysan5** [reply]
Kept the location search internal, for consistence with bone indices and weights

**Reply by raysan5** [reply]
raylib defaults to CPU-skinning, GPU-skinning must be enabled on raylib config with `RL_SUPPORT_MESH_GPU_SKINNING`.

**Reply by dc1248** [reply]
Looks about right. Should be very helpful for my game
I plan to test when merged and report on any issues I encounter.

Quick suggestions:
- Keep Transform* (instead of AnimPose typedef) for pointer clarity.
- Add 'name' to Model and Mesh; use for all imports that provide names.

**Reply by raysan5** [reply]
@dc1248 I've kept working on this for a while and I'm defining `ModelAnimPose` for convenience but no function exposes it, only kept internal to `Model`, `ModelSkeleton` and `ModelAnimation`. About adding mode `name` fields, structures are already quite big, I'm afraid those fields will impact performace and not that useful for most users (like animation or bones names).

**Reply by Kirandeep-Singh-Khehra** [reply]
Hi Ray, are there any plans to have split body animation and additive blending.

**Reply by raysan5** [reply]
I think current redesign is ok for now, it was also added a new example [`models_animation_blend_custom`](https://github.com/raysan5/raylib/blob/master/examples/models/models_animation_blend_custom.c) for users with more advance needs and I left room for a `UpdateModelAnimationPro()` version supporting more advance features.

**Reply by dc1248** [reply]
@raysan5 Fair enough. Thanks for doing all this btw!

As for Mesh and Material names (sorry, not Model) I **need** them in my model loading process, as identifiers and for suffix handling e.g. -col -navmesh etc. This is a common/standard convention, although it's mostly outside the scope of Raylib. Unfortunately LoadModel() throws the names away. Modding Raylib and sticking the names in Mesh+Material was the simplest temporary solution for me.

I suppose the proper solution is `LoadModelPro(const char *fileName, ModelMetadata *metadata)` where ModelMetadata is a struct with string tables to be allocated and filled by LoadGLTF() etc. This struct could be a catch-all for any other potentially useful metadata not needed in core functions. Thoughts?

**Comment by raysan5** [comment]
UPDATE: Examples involving models animations require a complete review, there are some from several contributors, I'll try to review all them, maybe some get deleted and some completely redesigned. Here a list with the examples I'd like to have:

 - `models_anim_playing` - Models loading with animation, multiple formats (IQM/M3D/GLTF), selectable animation, timeline (keyfframes)
 - `models_anim_timming` - GLTF model animations playing, timeline, playing speed can be adjusted, using frames fractional blending
 - `models_anim_pose_blending` - Blending between two animation poses, selectable from two UI-timelines (2 animations)
 - `models_anim_smooth_transition` - Two channels animations smooth transition, implemented on user side using the new API (it should allow it)

Also reviewing examples related to a specific models format (`_loading_gltf`, `_loading_m3d`, `_loading_iqm`), I prefer to keep the examples more generic and not specific to source data file formats.

`raygui` will be used on those examples for convenience.

**Comment by raysan5** [comment]
UPDATE: Redesign is finally done: https://github.com/raysan5/raylib/commit/d4dc038e2e77ad8b0fddcfa323ee872cc2eb1d0b

 - REVIEWED: Reorganized structures for a clearer distinction between "skeleton", "skin" and "skinning" data
 - ADDED: New structures: `ModelSkeleton`, `ModelAnimPose` (alias `Transform*`)
 - ADDED: Runtime data `currentPose` and `boneMatrices` to `Model` structure
 - ADDED: Support animation frames-blending, for timing control
 - ADDED: Support animations blending, between two animations
 - REVIEWED: All models animation loading functions
 - ADDED: `UpdateModelAnimationEx()` for two animations blending
 - REMOVED: `UpdateModelAnimationBones*()`, simplified API
 - REVIEWED: Shader attributes/uniforms names for animations, for consistency
 - REVIEWED: Multiple tweaks on animations loading for consistency between formats
 - ADDED: example: `models_animation_timing`
 - ADDED: example: `models_animation_blending`
 - REVIEWED: example: `models_animation_gpu_skinning`
 - REVIEWED: example: `models_animation_blend_custom`
 - REVIEWED: All animated models loading examples

**Reply by jestarray** [reply]
`UnloadModelAnimation` also got removed, but `UnloadModelAnimations`(plural) still exists. What did it get replaced with? I don't do any model stuff but I'm trying to update the bindings here

https://github.com/raysan5/raylib/commit/d4dc038e2e77ad8b0fddcfa323ee872cc2eb1d0b#diff-008ec254b9e10087d979356eaa799f93f24b3173a6a0c44445c24fee791a1e90L1626

**Reply by raysan5** [reply]
`UnloadModelAnimations()` is aligned to `LoadModelAnimations()`, never existed a `LoadModelAnimation()` so the inconsistency was just removed.

**Comment by dc1248** [comment]
@raysan5 I've beta tested the new system. I have a few minor issues I'll report here. (If you want me to open any separate issues or PRs, let me know.)

I'm using a custom version of `UpdateModelAnimationEx` as in [`models_animation_blend_custom`](https://github.com/raysan5/raylib/blob/master/examples/models/models_animation_blend_custom.c) for selective blending, jigglebones, and maybe IK in the future. I like this approach.

For CPU skinning, you could call `UpdateModelAnimationVertexBuffers` from `DrawModelEx` instead of from `UpdateModelAnimation/Ex`.
This eliminates the need to duplicate it in `models_animation_blend_custom.c` and allows additional bone manipulation after `UpdateModelAnimation` (which may be useful even when using a custom variant).
Optimization: add a `Model.vertexBufferUpdateRequired` flag (set by `UpdateModelAnimation/Ex`). There are fairly common use cases for this, e.g. zombie hordes that update at reduced frequency when they're on the periphery of the visible area.

GPU SKINNING
1. For troubleshooting and optimization, it would be nice if I could selectively enable CPU and/or GPU skinning at runtime (when raylib is compiled with `SUPPORT_GPU_SKINNING`). I'm thinking a global enum variable to enable CPU, GPU, BOTH, or NONE; this would enable uploading of vertices and/or boneMatrices in `DrawModelEx`.
2. In some cases it may be desirable to recompile shaders with or without skinning, but Raylib should not enforce or assume it.
3. Examples should say whether they're actually using CPU or GPU skinning.
4. I suggest adding `IsShaderGPUSkinning(shader)` which returns false if any of the shader attributes/uniforms are -1 (or if globally disabled).
5. An open question: Should the default shader should support GPU skinning?
    - PROS: "it just works", no need for custom shaders in basic animation examples, could be useful for troubleshooting.
    - CONS: complexity, reliability concerns, maybe performance.

**Comment by raysan5** [comment]
> For CPU skinning, you could call `UpdateModelAnimationVertexBuffers` from `DrawModelEx` instead of from `UpdateModelAnimation/Ex`.
This eliminates the need to duplicate it in `models_animation_blend_custom.c` and allows additional bone manipulation after `UpdateModelAnimation` (which may be useful even when using a custom variant).

I prefer current approach, so the vertex buffers are only updated when there is an animation update and animation logic is localized in a single function, if placed on `DrawModelEx()` it will update always (with a performance hit) or it will require managing an additional flag internally (that I prefer to avoid for simplicity). The `models_animation_blend_custom.c` example is provided to give advance users that flexibility when required.

1. That's something I though a lot, if both options should be available, but I can't really think a situation where a user using GPU-skinning could go for CPU-skinning, actually, also Unity had both option separate and required enabling GPU-skinning when desired. The cost of keeping and updating those extra CPU buffers could be high. Also, raylib prioritizes simplicity, always, managing additional flags and configuration options to allow both approaches adds complexity to the system.
2. raylib tries to find some default locations for you on shader, if you prefer not to use it you can just rename those variables in your shaders and they won't be set.
3. There are some comments on examples, actually they are designed to support both approaches, depending how raylib has been compiled but I agree some log info on raylib initialization could be useful, I will add it.
4. It can be checked at user side, no need for an additional API function.
5. I don't think so, default shaders intent to be as minimal as possible, most raylib users only need them for 2D.

**Reply by dc1248** [reply]
Okay, in that case I think you should expose 3 separate functions:
`UpdateModelAnimationBones()` - for all model animations
`UpdateModelAnimationVertexBuffers()` - for CPU skinning only
`UpdateModelAnimation()` - a wrapper for both of the above

I think I understand your roadmap now: 2D-focused with solid foundations for 3D, with a rmodels.c and a nice collection of examples to test the foundations, but you want to leave more advanced 3D to third-party libraries like https://github.com/Bigfoot71/r3d, correct? That seems reasonable and maintainable.

**Reply by raysan5** [reply]
Thanks but I can't see the need for those additional API function, I prefer to keep the API minimal and let advance users implement any additional function needed, after all, data structures are transparent to allow that.
